provider/kv/kv_config.go

```diff
@@ -575,7 +575,11 @@ func (p *Provider) listServers(backend string) []string {
 
 func (p *Provider) serverFilter(serverName string) bool {
 	key := fmt.Sprint(serverName, pathBackendServerURL)
-	if _, err := p.kvClient.Get(key, nil); err != nil {
+	if p.CacheKeyValues && p.kvCache != nil {
```

Comment:
there are two things that bother me at every place you do a cache read in this patch:

1) at every cache miss, you consider that the value does not actually exist at all. But the way a cache works is that usually you do

```
val, ok := cache.get(key)
if !ok {
  // cache miss
  val, ok = upstreamSource.get(key)
  if !ok {
    return errors.New("not found")
  }
}
```

So I assume that you took this approach because you think the whole world is in cache, so when we do a get on the cache, it's equivalent to doing a get on the upstream source. But are we sure that is true?

2) I think it would be more readable if we had a Getter on the provider, that is in charge of abstracting away the cache read and the upstream source read, i.e. something like:

```
func (p *Provider) Get(key string) (string, err) {
  if !p.CacheKeyValues || p.kvCache == nil {
    return p.kvClient.Get(key, nil)
  }
  val, ok := p.kvCache[key]
  if !ok {
    return p.kvClient.Get(key, nil)
  }
}
```

I know the way we treat errors or values not found is a bit different at each place in the code, but maybe there's still a way to abstract away some of the repetition?